lib/rules/no-return-await.js

```diff
@@ -27,6 +27,46 @@ module.exports = {
     create(context) {
 
         /**
+         * Reports a found unnecessary `await` expression.
+         * @param {ASTNode} node The node representing the `await` expression to report
+         * @returns {void}
+         */
+        function reportUnnecessaryAwait(node) {
+            context.report({
+                node: context.getSourceCode().getFirstToken(node),
+                loc: node.loc,
+                message,
+            });
+        }
+
+        /**
+        * Finds recursively an `await` expression that is a possible resultant return and is considered unnecessary by the definition of this rule.
+        * This recursion is necessary to cover the cases which `return` argument is formed by comma sequences and/or ternary conditions that could result in an `await` expression.
+        * @param {ASTNode} node A node that is the `return` argument or decendent
+        * @returns {void}
+        */
+        function findUnnecessaryAwait(node) {
+            switch (node.type) {
+                case "AwaitExpression":
+                    reportUnnecessaryAwait(node);
+                    break;
+
+                case "SequenceExpression":
+                    findUnnecessaryAwait(node.expressions[node.expressions.length - 1]);
```

Comment:
We try to avoid recursing down the AST within rules, instead relying on the built-in traversal that core is already doing.

Conceptually, whenever we detect that traversal has entered a `return` inside an `async` function, set a flag, and clear that flag on exit from the same `ReturnStatement`. Whenever we encounter an `AwaitExpression`, if the flag is set, report an error.

That works for cases like

```js
async function foo() {
    return a() ? await b() : c();
}
```

But there are edge cases where that would get tripped up:

```js
async function foo() {
    return async function () {
        const bar = await baz();
        // ...
    };
}
```

To avoid that case, we need to maintain a stack that separates function scopes.

1. When traversal enters a function `push` an unset flag onto the stack.
1. When traversal enters a `ReturnStatement`, set the topmost flag on the stack.
1. When traversal enters an `AwaitExpression`, if the topmost flag on the stack is set, warn because this `await` is inside a `return`.
1. When traversal leaves a `ReturnStatement`, unset the topmost flag on the stack.
1. When traversal leaves a function, pop the topmost flag off of the stack.